rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,222 @@
+---
+authors: Noah Stride (noah.stride@goteleport.com)
+state: draft
+---
+
+# RFD 00143 - External Kubernetes delegated joining (jwks)
+
+## Required approvers
+
+* Engineering: @zmb3 && (@hugoShaka || @tigrato)
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What
+
+This RFD proposes a method for allowing entities (such as a Machine ID instance
+or an Agent) to join Teleport Clusters via federation between the Kubernetes 
+Cluster they reside in and a Teleport Cluster, where that Teleport Cluster does 
+not reside in the same Kubernetes Cluster.
+
+This is distinct from 
+[RFD 0094 - Kubernetes node joining](./0094-kubernetes-node-joining.md) which
+only supports the joining entity existing within the same Kubernetes Cluster
+as the Teleport Cluster.
+
+## Why
+
+It is not unusual for users to wish to deploy Machine ID bots and Teleport
+Agents in Kubernetes Clusters which is not the same cluster in which their
+Auth Service resides. For example, they may operate many Kubernetes Clusters or
+they may use Teleport Cloud (in which we host their Teleport Cluster).
+
+Currently, these users must use a different join method, and, for users
+deploying into environments where there is no platform-specific delegated
+join method, they must fallback to using the `token` join method which is a
+sensitive, long-lived, secret. This also requires some element of state, and
+is inconvenient in more ephemeral use-cases.
+
+To provide a more concrete use-case, many users currently deploy Teleport
+Plugins (such as the Slack Access Request plugin). Adding support for External
+Kubernetes Joining will provide a golden pathway for this deployment using
+Machine ID that avoids long-lived shared secrets and provides a seamless 
+experience.
+
+## Details
+
+### Configuration
+
+This join method will be added to the existing `ProvisionTokenV2` specification.
+
+In order to clearly distinguish this new join method from the existing
+`kubernetes` join method, this method will be named `kubernetes-jwks`.
+
+Similar to other join methods, users will be able to configure a list of allow
+rules. If a joining entities presented token matches any of the allow rules 
+then it will be allowed to join.
+
+Each join token will be linked to a specific Kubernetes Cluster. It will not
+be possible to create a join token which can be used for joining from multiple
+Kubernetes Clusters.
+
+Example token configuration:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: kubernetes-jwks-token
+spec:
+  roles: ["Bot"]
+  bot_name: argocd
+  join_method: "kubernetes-jwks",
+  kubernetes_jwks:
+    # jwks is obtained by the user by following the steps after this example
+    # configuration.
+    jwks: |
```

Comment:
One of my concerns with accepting multiple JWKs is it make be difficult to determine which belongs to which of your clusters :/

Reply:
I added the `cluster_name` to solve that problem

Reply:
Are you proposing that the allow rules are per each cluster, or shared between all the clusters ? Your example YAML is a little mangled.

Reply:
true.

Sorry about that 

The real form is to allow specifying a jwks payload and allow arms per cluster but the token supports multiple clusters.

```yaml
kind: token
version: v2
metadata:
  name: kubernetes-jwks-token
spec:
  roles: ["Bot"]
  bot_name: argocd
  join_method: "kubernetes-jwks"
  kubernetes_jwks:
   - cluster_name: cluster1
     allow:
       - service_account: "my-namespace:my-service-account"
       - service_account: "my-namespace:my-service-account"
     jwks: |
       {"keys":[{"use":"sig","kty":"RSA","kid":"","alg":"RS256","n":"","e":"AQAB"}]}
   - cluster_name: cluster2
     allow:
       - service_account: "my-namespace:my-service-account"
       - service_account: "my-namespace:my-service-account"
     jwks: |
       {"keys":[{"use":"sig","kty":"RSA","kid":"","alg":"RS256","n":"","e":"AQAB"}]}


```

Go representation

```go

type KubeJWKSEntry struct{
   JWKS string
   ClusterName string
   Allow Allow
}

types Token struct{
  JoinMethod string
  KubernetesJWKS []KubeJWKSEntry
}


```

Reply:
Would you expect the joining client to specify which "cluster" it is joining as, or would this be inferred from a matching JWKS ?

Reply:
Cluster is inferred from JWKS but it's only used to identify which cluster the JWKS belongs to in case they rotate the certificate, it's easier to know what to replace

Reply:
I am ok with it, although I think allowing some kind of service account restriction per cluster would be beneficial

Reply:
Mhm.

We could potentially allow an `allow` case to be scoped to a specific cluster. e.g

```
kubernetes_jwks:
  clusters:
    - name: cluster-1
      static_jwks: {"keys":[{"use":"sig","kty":"RSA","kid":"","alg":"RS256","n":"","e":"AQAB"}]}
    - name: cluster-2
      static_jwks: {"keys":[{"use":"sig","kty":"RSA","kid":"","alg":"RS256","n":"","e":"AQAB"}]}
  allow:
   # This service_account can only be used in cluster-1
    - service_account: my-namespace:my-service-account
      cluster: cluster-1
   # This service_account can only be used in any cluster.
    - service_account: my-namespace:my-service-account
```

Although - I worry that this turns the name we were using purely for "descriptive" purposes into something thats actually going to be used as part of a criteria.

Reply:
[b2826d4](https://github.com/gravitational/teleport/pull/29225/commits/b2826d48c4fd5b833955e3f3af93647a0c4115ba)